### PR TITLE
neovim: new package at 0.4.2

### PR DESCRIPTION
**Pull request by WillEccles**
## Description of package
Neovim: hyper-extensible Vim-based text editor. This is version 0.4.2, which is the latest stable version.
## Checklist

- [x] Latest upstream version.
- [x] I agree to maintain this package (**required**).
- [x] I agree to notify KISS if I can no longer maintain this package (**required**).

Note: I'm not certain that I can maintain this package consistently. May not be able to do this in the future.

### CODE REVIEW

**Comment by dylanaraps** [comment]
I'm assuming that this includes all dependencies vendored (*`libtermkey`, `libuv`, `libvterm`, `msgpack` and even LuaJIT*)? Interesting, I wasn't aware that this could be done with `neovim`. From what I can see in this PR only `cmake` is needed as a dependency?

File: community/neovim/depends [hunk 1]
```diff
@@ -0,0 +1 @@
+cmake
```
Comment (dylanaraps) [inline community/neovim/depends]: Should be `cmake make` as I assume `cmake` isn't needed at runtime (*only at compile time*).

File: community/neovim/build [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh -e
+
+make CMAKE_BUILD_TYPE=Release
+
```
Comment (dylanaraps) [inline community/neovim/build]: This blank line can go.

**Comment by WillEccles** [comment]
> I'm assuming that this includes all dependencies vendored (_`libtermkey`, `libuv`, `libvterm`, `msgpack` and even LuaJIT_)? Interesting, I wasn't aware that this could be done with `neovim`. From what I can see in this PR only `cmake` is needed as a dependency?

Yep. Neovim uses CMake to do dependencies all by itself. I was surprised as well. You can, of course, manually install all of them, but it can just download and build them as needed if you don't want to do that.

File: community/neovim/build (hunk 2 above)
Comment (WillEccles) [reply community/neovim/build]: Removed.

File: community/neovim/depends (hunk 1 above)
Comment (WillEccles) [reply community/neovim/depends]: Fixed in f19fb7f

**Comment by WillEccles** [comment]
Did not mean to mark those as resolved.

Fixed reviews in f19fb7f. Recommend squashing.

**Comment by dylanaraps** [comment]
> but it can just download and build them as needed if you don't want to do that.

Ah, my question was whether or not it included them in the source tarball itself. I'm against packages starting downloads during a build for obvious reasons. Exceptions have been made of course and I need to think about that here.

From what I can conclude the build will download and compile the dependencies and then statically link them into the resulting `neovim` binary? At the very least the resulting `neovim` will be *partially* static (*only the external sources will be statically linked*)?

Interesting.

**Comment by WillEccles** [comment]
It will be statically linked with the dependencies. However, the dependencies are not included in the source tarball. I would say that for a well-known and oft-built piece of software such as nvim, downloading deps at build-time is probably fine, but that's just my 2¢.

**Comment by WillEccles** [comment]
How does `cmake make` work, by the way? Is `make` just a keyword, or is this a fancy way of saying that `make` requires `cmake` when building or something?

**Comment by dylanaraps** [comment]
The reason I'm reluctant is that checksums and signatures against whatever `neovim`'s build system downloads can't be verified. It may be possible to extract the download URLS, use them as regular sources in the `kiss` package and `cp` them to the right location but there would have to be support from upstream for this kind of build.

As it stands in KISS there are only two packages which download during build, `mpv` and `rust`. I'm working locally on removing the network requirement from both of these which will make all sources 100% verifiable. (*This is another reason why I'd rather not grow this list*).

Also for reference: The possible solution for `neovim` is exactly how I'm fixing `rust`'s build.

> How does cmake make work, by the way? Is make just a keyword, or is this a fancy way of saying that make requires cmake when building or something?

In `neovim`'s case a `Makefile` is provided as a wrapper around `cmake` "for convenience" though to me it just seems like an abstraction over the "real" build system and an additional dependency on `make` (*though `make` should be available regardless*).

**Comment by WillEccles** [comment]
Unfortunately, I do not have a system at this time which I can use to test dependencies, so for the time being I will have to leave this one as-is. I may try to install KISS in a VM at some point, since I already have two OSes on this system and don't want to partition for a third. Then I would be able to fix this.

**Comment by konimex** [comment]
> How does `cmake make` work, by the way? Is `make` just a keyword, or is this a fancy way of saying that `make` requires `cmake` when building or something? I assume the former.

If you're talking about the `depends` file, `cmake make` is just a way of telling the package manager that the `cmake` package is only used to build `neovim` and can be safely uninstalled if `neovim` has been installed.

**Comment by WillEccles** [comment]
@konimex Thanks, that's about how I figured it worked.

**Comment by dylanaraps** [comment]
Neovim supports caching the build directory: https://github.com/neovim/neovim/blob/master/third-party/cmake/DownloadAndExtractFile.cmake#L26-L37

```cmake
# Check whether the source has been downloaded. If true, skip it.
# Useful for external downloads like homebrew.
if(USE_EXISTING_SRC_DIR)
  if(EXISTS "${SRC_DIR}" AND IS_DIRECTORY "${SRC_DIR}")
    file(GLOB EXISTED_FILES "${SRC_DIR}/*")
    if(EXISTED_FILES)
      message(STATUS "${SRC_DIR} is found and not empty, skipping download and extraction. ")
      return()
    endif()
  endif()
  message(FATAL_ERROR "USE_EXISTING_SRC_DIR set to ON, but '${SRC_DIR}' does not exist or is empty.")
endif()```

File: community/neovim/depends [hunk 3]
```diff
@@ -0,0 +1 @@
+cmake make
```
Comment (dylanaraps) [inline community/neovim/depends]: Also missing a depend on `libtool`.

(`libtool make`)

Comment (dylanaraps) [reply community/neovim/depends]: Also missing `automake make`.

**Comment by dylanaraps** [comment]
With the dependency fixes the build still fails due to the absence of `gettext`. Has this been tested on your machine by any chance?

**Comment by WillEccles** [comment]
This functioned on my machine, although I am not running KISS, just the package manager. My apologies if I missed a few dependencies. I may have already had more things installed than I had thought, likely as dependencies for other things.

**Comment by WillEccles** [comment]
I plan to get KISS running in a VM sometime in the near future to make sure it's worth replacing my current install with (because work-related things are important to keep working). At that point I will be able to fully test and flesh out this package.

**Comment by dylanaraps** [comment]
This fixes the `gettext` issue:

```sh
# Remove mandatory gettext dependency.
sed -i -e 's/^/#/' src/nvim/po/CMakeLists.txt
```

**Comment by WillEccles** [comment]
Interesting. Is this because vim (and nvim by extension, I presume) does not require gettext? I assume nvim just made it "mandatory" because they just figured everyone would have it.

**Comment by WillEccles** [comment]
Also, does it build now? I was installing KISS on a VM earlier and ran into a rather concerning issue installing GRUB (it complained about ext2, but it was on an ext4 partition). Otherwise I would have tested by now.

File: community/neovim/depends (hunk 3 above)
Comment (WillEccles) [reply community/neovim/depends]: Addressed in https://github.com/kisslinux/community/pull/49/commits/2d67047abee64fd39d7a82ace7d3092356354c69

File: community/neovim/build [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh -e
+
+# Remove mandatory gettext dependency.
+sed -i -e 's/^/#/' src/nvim/po/CMakeLists.txt
```
Comment (dylanaraps) [inline community/neovim/build]: Add a newline below the `sed` line.

File: community/neovim/build [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+make CMAKE_BUILD_TYPE=Release
+make CMAKE_INSTALL_PREFIX="$1" install
```
Comment (dylanaraps) [inline community/neovim/build]: Change `CMAKE_INSTALL_PREFIX` to `DESTDIR`.

**Comment by dylanaraps** [comment]
Also, @dithpri is working on removing the network requirement during build time. :+1:

**Comment by WillEccles** [comment]
Cool, thanks!

**Comment by WillEccles** [comment]
Fixed `build` in https://github.com/kisslinux/community/pull/49/commits/a8e77dd0ce4e8d1a0c734ef7e6fbdb3f49226adf.

**Comment by dithpri** [comment]
Partly done caching, cf. [my neovim branch](https://github.com/dithpri/community/tree/neovim/community/neovim)

**Comment by dylanaraps** [comment]
@dithpri this should suffice for the dependencies:

`sources`: 

```
https://github.com/libuv/libuv/archive/v1.30.0.tar.gz?noextract-libuv .deps/build/src/libuv/
```

The second field in `sources` specifies the extraction directory and it is created if it doesn't exist. The directory is relative to the build directory.

**Comment by dylanaraps** [comment]
The stance on network connection in `community` has been relaxed and builds like this are clear now. Merging this. I'll maintain it until someone else wishes to take over. :)

**Pull request merged**